### PR TITLE
Add a Facebook event konnector, which fetch user's event through FB OAuth2 API

### PR DESCRIPTION
**Pull request by jacquarg**
A Konnector which add "accepted" Facebook event to a Calendar in the Cozy.
This Konnector is mainly a proof of concept of using OAuth2 API with a Cozy (which  should open the door to about all modern APIs), in a confident way. Here, the Konnector use a _redirect_uri proxy_ which is a simple page, hosted on github.io : https://github.com/jacquarg/jacquarg.github.io/blob/master/proxy_redirect/facebook_events/index.html

Actually, the Konnector requires to Copy/Past the token ; but everything is setup for automatic handling (with redirections, ...) - it's not done yet because Konnector UI don't provide easy access to the UI and routes.

### CODE REVIEW

**Comment by ZeHiro** [comment]
HI @jacquarg could you lint the code please, by running `npm run lint` and correct all the errors? Thank you.

**Comment by frankrousseau** [comment]
@ZeHiro what do you think about merging this PR? I mean, the fact that it doesn't work will problem annoy people. It could motivate them to fix the konnector. If noone complains about the fact it fails, will be a proof that noone cares about this konnector and we will be able to remove it.

**Comment by ZeHiro** [comment]
@frankrousseau as the PR is on the dev branch, it is not a problem, it won't be used :-b . 
But I quite don't like publishing smth that appears not to work (at least if it is a systematic fail such as it is here). The rest of my comments are just quality improvements.

**Comment by jacquarg** [comment]
@ZeHiro, I can find some time to do many of the improvements you asked. But I would secure the biggest problem first ; which that it looks to not work with your account.

**Comment by ZeHiro** [comment]
@jacquarg 
I do have such events. I'll retry tonight on my dev machine.

**Comment by ZeHiro** [comment]
@jacquarg 
I get a 200 answer : 
{"data":[]}
But with de dev interface of facebook, I get the events. 
In the fb parameters of the app, I get this app can "Vous envoyer des notifications". Did you set the approripate rights for the app?

File: server/konnectors/facebook_events.js [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+      })[0];
+
+      currentAccount.accessToken = data.accessToken;
+
+    } catch (e) {
+      connector.logger.error("Can't fetch konnector instances");
+      return callback(e);
+    }
+
+    konnector.updateAttributes({ accounts: konnector.accounts }, callback);
```
Comment (ZeHiro) [inline server/konnectors/facebook_events.js]: Can you catch the error in the updatr of the konnector?

**Comment by jacquarg** [comment]
Thanks for your feedbacks. I reproduce the _non functionning_ bug. 
I didn't visualize it before because my Facebook account is the developer account of the Facebook app. So the state is, the app has to be _verified_ before facebook allows it to fetch user_events. Facebook claims that its a manual procedure, and that I have to explain the steps to use the app and connect to Facebook. 
@frankrousseau, do I have to take some precautions before speaking about Cozy to some Facebook team buddies ; ) ?

**Comment by ZeHiro** [comment]
Maybe should you already request for the contact rights, so you don't have to pass the test twice

**Comment by frankrousseau** [comment]
@jacquarg You can speak to them about Cozy. The problem is more for Facebook, their employees could be tempted to join us ;).

**Comment by jacquarg** [comment]
Just update the Facebook form with more info and a screencast. They (a human) should look at it in 5 business days, we will see...
But if it fails with Facebook, we may try with other OAut2 providers, with simpler submission procedure (Googe, LinkedIn, Instagram, ...)

**Comment by jacquarg** [comment]
Good news, Facebook approve the apps for events. 
I updated the konnector with the new appId. You could try again, it should work now.

I try to apply some of the fixes you recomand ; but the lint is pushy with ES !

File: client/app/locales/fr.coffee [hunk 2]
```diff
@@ -159,6 +160,7 @@ module.exports =
     'notification sfr_mobile': "%{smart_count} nouvelle facture importée |||| %{smart_count} nouvelles factures importées"
     'notification doctolib creation': "%{smart_count} nouveau rendez-vous importé |||| %{smart_count} nouveaux rendez-vous importés"
     'notification doctolib update': "%{smart_count} rendez-vous mis à jour |||| %{smart_count} rendez-vous mis à jour"
+    'notification facebook_events': "%{smart_count} événement mis à jour. |||| %{smart_count} événements mis à jour."
```
Comment (ZeHiro) [inline client/app/locales/fr.coffee]: should be 'notification facebook_events update'

'notification facebook_events creation' is missing (see the function buildNotifContent)

File: server/konnectors/facebook_events.js [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+      }
+      next(err);
+    });
+}
+
+// Save konnector's fieldValues during fetch process.
+function saveTokenInKonnector(requiredFields, entries, data, callback) {
+  connector.logger.info('Save refreshed token.');
+
+  const Konnector = require('../models/konnector');
```
Comment (ZeHiro) [inline server/konnectors/facebook_events.js]: should be at the top of the file.

File: server/konnectors/facebook_events.js [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+  // Konnector.get(connector.slug, function(err, konnector) {
+  Konnector.all((err, konnectors) => {
+    if (err) {
+      connector.logger.error("Can't fetch konnector instances", + err.msg);
+      return callback(err);
+    }
+    let konnector = null;
+    try {
+      konnector = konnectors.filter((k) => {
+        return k.slug === connector.slug;
```
Comment (ZeHiro) [inline server/konnectors/facebook_events.js]: The `{ } ` are not required, as well as the  `return` statement

File: server/konnectors/facebook_events.js [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+      })[0];
+
+      // Find which account we are using now.
+      const currentAccount = konnector.accounts.filter((account) => {
```
Comment (ZeHiro) [inline server/konnectors/facebook_events.js]: Same as previous comment

File: server/konnectors/facebook_events.js [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+          lastUpdate: new Date().toISOString(),
+        },
+      };
+    } catch (e) {
+      connector.logger.error('Skip an event while parsing it.', e);
+      return null;
+    }
+  });
+
+  entries.events = list.filter(ev => { return ev !== null; });
```
Comment (ZeHiro) [inline server/konnectors/facebook_events.js]: The `{ }` are not required, as well as the `return` statement

**Comment by ZeHiro** [comment]
@jacquarg 
Works well now, there still are some comments directly in the code. Thank you, i'll then merge the PR.

File: server/konnectors/facebook_events.js [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+      const date = new Date(fbEvent.start_time);
+
+      let locationStr = '';
+      if (fbEvent.place) {
+        locationStr = fbEvent.place.name;
+        if (fbEvent.place.location) {
+          locationStr += `, ${fbEvent.place.location.street}, `;
+          locationStr += fbEvent.place.location.city;
+          locationStr += ' ';
+          locationStr += fbEvent.place.location.zip;
```
Comment (ZeHiro) [inline server/konnectors/facebook_events.js]: This one can be absent

File: server/konnectors/facebook_events.js (hunk 3 above)
Comment (jacquarg) [inline server/konnectors/facebook_events.js]: Konnector is empty ( `{}` ) if require is moved to the top. The reason is : 
1. [Konnector requires konnector_hash](https://github.com/cozy-labs/konnectors/blob/master/server/models/konnector.coffee#L3)
2. [Which load each konnectors modules](https://github.com/cozy-labs/konnectors/blob/master/server/lib/konnector_hash.coffee#L22-L26) 
So, if we requires Konnector in the body of a konnector, we get the Konnector at [this step](https://github.com/cozy-labs/konnectors/blob/master/server/models/konnector.coffee#L3), which is `{}`

Comment (ZeHiro) [inline server/konnectors/facebook_events.js]: Makes sense

**Comment by ZeHiro** [comment]
@jacquarg final remarque, the tests do not pass because you have left `;` in the arrow functions. Can you fix this please?

**Comment by frankrousseau** [comment]
Suggestion: when minor fixes are required to merge a pull request. What about merging the PR and doing the fix ourselves (as maintainers)?
It would make things smoother.

**Comment by ZeHiro** [comment]
@frankrousseau Good idea. I'll do like that. Thanks @jacquarg , I take care of all the fixes.

**Comment by ZeHiro** [comment]
@jacquarg, @frankrousseau   Can I merge directly to master? (see #303)

**Comment by frankrousseau** [comment]
Yes @ZeHiro

**Comment by jacquarg** [comment]
Yes @ZeHiro ! Would be great.
If there is some enthusiasm about it, we may think about a clean UI dedicated to OAuth 2 ( whithout copy/paste) and create Facebook Contact, maybe google, linked in, and others connectors like that !

**Comment by ZeHiro** [comment]
Ok, the i close.

**Pull request closed without merging**